File: spring/src/test/java/com/hcsy/spring/common/utils/InternalTokenUtilTest.java
package com.hcsy.spring.common.utils;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import com.hcsy.spring.common.exceptions.BusinessException;
import com.hcsy.spring.core.properties.InternalTokenProperties;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mock;

class InternalTokenUtilTest {

    private static final String SECRET = "unit-test-internal-token-secret-32-bytes";
    private InternalTokenUtil tokenUtil;

    @BeforeEach
    void setUp() {
        tokenUtil = createTokenUtil(SECRET);
    }

    // 验证该场景的预期行为

    @Test
    @DisplayName("生成并解析内部令牌声明")
    void generatesAndExtractsInternalTokenClaims() {
        InternalTokenUtil configuredTokenUtil = createTokenUtil(resolveConfiguredSecret());
        String token = configuredTokenUtil.generateInternalToken(10001L, "spring");

        System.out.println("生成的内部Token: " + token);

        assertTrue(configuredTokenUtil.validateInternalToken(token));
        assertEquals(10001L, configuredTokenUtil.extractUserId(token));
        assertEquals("spring", configuredTokenUtil.extractServiceName(token));
        assertTrue(configuredTokenUtil.getRemainingTime(token) > 0);
    }

    // 验证该场景的预期行为

    @Test
    @DisplayName("拒绝使用其他密钥签名的内部令牌")
    void rejectsTokenSignedWithAnotherSecret() {
        String token = tokenUtil.generateInternalToken(10001L, "spring");
        InternalTokenUtil verifier = createTokenUtil("another-unit-test-secret-with-32-bytes");

        assertThrows(BusinessException.class, () -> verifier.validateInternalToken(token));
    }

    // 验证该场景的预期行为

    @Test
    @DisplayName("拒绝格式错误的内部令牌")
    void rejectsMalformedToken() {
        assertThrows(BusinessException.class, () -> tokenUtil.validateInternalToken("not-a-jwt"));
    }

    private InternalTokenUtil createTokenUtil(String secret) {
        InternalTokenUtil util = new InternalTokenUtil(
            new InternalTokenProperties(secret, 60_000L),
            mock(SimpleLogger.class));
        util.initKey();
        return util;
    }

    private String resolveConfiguredSecret() {
        String environmentSecret = System.getenv("INTERNAL_TOKEN_SECRET");
        if (environmentSecret != null && !environmentSecret.isBlank()) {
            return environmentSecret.trim();
        }

        String systemSecret = System.getProperty("INTERNAL_TOKEN_SECRET");
        if (systemSecret != null && !systemSecret.isBlank()) {
            return systemSecret.trim();
        }

        for (Path candidate : List.of(Path.of(".env"), Path.of("spring/.env"), Path.of("../.env"))) {
            String fileSecret = readSecret(candidate);
            if (fileSecret != null) {
                return fileSecret;
            }
        }
        return SECRET;
    }

    private String readSecret(Path path) {
        if (!Files.exists(path)) {
            return null;
        }
        try {
            return Files.readAllLines(path).stream()
                .map(String::trim)
                .filter(line -> line.startsWith("INTERNAL_TOKEN_SECRET="))
                .map(line -> stripQuotes(line.substring(line.indexOf('=') + 1).trim()))
                .filter(value -> !value.isBlank())
                .findFirst()
                .orElse(null);
        } catch (IOException e) {
            return null;
        }
    }

    private String stripQuotes(String value) {
        if (value.length() >= 2 &&
            ((value.startsWith("\"") && value.endsWith("\"")) ||
                (value.startsWith("'") && value.endsWith("'")))) {
            return value.substring(1, value.length() - 1);
        }
        return value;
    }
}

File: gozero/app/common/utils/internalTokenUtil_test.go
package utils_test

import (
	"fmt"
	"os"
	"path/filepath"
	"testing"

	"app/common/utils"

	"github.com/joho/godotenv"
)

const testInternalTokenSecret = "unit-test-internal-token-secret-32-bytes"

func TestInternalTokenRoundTrip(t *testing.T) {
	_ = godotenv.Load(filepath.Join("..", "..", ".env"))
	secret := os.Getenv("INTERNAL_TOKEN_SECRET")
	if secret == "" {
		secret = testInternalTokenSecret
	}
	if err := utils.InitInternalTokenUtil(secret, 60000); err != nil {
		t.Fatalf("初始化内部令牌工具失败: %v", err)
	}
	tokenUtil, err := utils.GetTokenUtil()
	if err != nil {
		t.Fatalf("获取内部令牌工具失败: %v", err)
	}

	token, err := tokenUtil.GenerateInternalToken(10001, "gozero")
	if err != nil {
		t.Fatalf("生成内部令牌失败: %v", err)
	}
	fmt.Printf("生成的内部Token: %s\n", token)
	claims, err := tokenUtil.ValidateInternalToken(token)
	if err != nil {
		t.Fatalf("验证内部令牌失败: %v", err)
	}
	if claims.ExtractUserID() != 10001 {
		t.Errorf("UserID = %d, 期望 10001", claims.ExtractUserID())
	}
	if claims.ExtractServiceName() != "gozero" {
		t.Errorf("ServiceName = %q, 期望 gozero", claims.ExtractServiceName())
	}
	if claims.TokenType != "internal" {
		t.Errorf("TokenType = %q, 期望 internal", claims.TokenType)
	}
}

func TestInternalTokenRejectsDifferentSecret(t *testing.T) {
	if err := utils.InitInternalTokenUtil(testInternalTokenSecret, 60000); err != nil {
		t.Fatalf("初始化签发工具失败: %v", err)
	}
	issuer, _ := utils.GetTokenUtil()
	token, err := issuer.GenerateInternalToken(10001, "gozero")
	if err != nil {
		t.Fatalf("生成内部令牌失败: %v", err)
	}

	if err := utils.InitInternalTokenUtil("another-unit-test-secret-with-32-bytes", 60000); err != nil {
		t.Fatalf("初始化验证工具失败: %v", err)
	}
	verifier, _ := utils.GetTokenUtil()
	if _, err := verifier.ValidateInternalToken(token); err == nil {
		t.Fatal("使用不同密钥签名的令牌应该验证失败")
	}
}

func TestInitInternalTokenUtilRejectsEmptySecret(t *testing.T) {
	if err := utils.InitInternalTokenUtil("", 60000); err == nil {
		t.Fatal("空密钥应该初始化失败")
	}
}
